src/java.base/share/classes/sun/security/provider/HSS.java

```diff
@@ -0,0 +1,744 @@
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package sun.security.provider;
+
+import sun.security.util.*;
+import sun.security.x509.AlgorithmId;
+import sun.security.x509.X509Key;
+
+import java.io.*;
+import java.security.*;
+import java.security.spec.*;
+import java.util.Arrays;
+import java.util.Objects;
+
+/*
+ * This class implements the Hierarchical Signature System using the
+ * Leighton-Micali Signatures (LMS) as described in RFC 8554 and NIST Special publication 800-208
+ */
+public final class HSS extends SignatureSpi {
+    private HSSPublicKey pubKey;
+    private ByteArrayOutputStream messageStream;
+
+    @Deprecated
+    protected void engineSetParameter(String param, Object value) {
+        throw new InvalidParameterException("No settable parameters exist for HSS/LMS");
+    }
+
+    @Deprecated
+    protected AlgorithmParameters engineGetParameter(String param) {
+        throw new InvalidParameterException("No parameters exist for HSS/LMS");
+    }
+
+    protected void engineInitSign(PrivateKey privateKey) throws InvalidKeyException {
+        throw new InvalidKeyException("HSS/LMS signing is not supported");
+    }
+
+    // This will never be called because engineInitSign unconditionally throws an exception
+    protected byte[] engineSign() throws SignatureException {
+        throw new SignatureException("HSS/LMS signing is not supported");
+    }
+
+    protected void engineInitVerify(PublicKey publicKey) throws InvalidKeyException {
+        HSSPublicKey pub;
+        if (publicKey instanceof HSSPublicKey p) {
+            pub = p;
+        } else {
+            KeyFactoryImpl factory = new HSS.KeyFactoryImpl();
+            Key pk = factory.engineTranslateKey(publicKey);
+            pub = (HSSPublicKey) pk;
+        }
+        pubKey = pub;
+        messageStream = new ByteArrayOutputStream();
+    }
+
+    protected void engineUpdate(byte data) {
+        messageStream.write(data);
+    }
+
+    protected void engineUpdate(byte[] data, int off, int len) {
+        messageStream.write(data, off, len);
+    }
+
+    protected boolean engineVerify(byte[] signature) throws SignatureException {
+        try {
+            HSSSignature sig = new HSSSignature(signature, pubKey);
+            LMSPublicKey lmsPubKey = pubKey.lmsPublicKey;
+            boolean result = true;
+            for (int i = 0; i < sig.Nspk; i++) {
+                byte[] keyArr = sig.pubList[i].keyArray();
+                result &= lmsVerify(lmsPubKey, sig.siglist[i], keyArr);
+                lmsPubKey = sig.pubList[i];
+            }
+
+            result &= lmsVerify(lmsPubKey, sig.siglist[sig.Nspk], messageStream.toByteArray());
+            return result;
+        } catch (Exception e) {
```

Comment:
Do we need this `catch`? I think only `SignatureException` can be thrown inside the `try` block.